pkg/merkle/tree.go

```diff
@@ -0,0 +1,320 @@
+	return m.tree
+}
+
+// Leaves returns the raw elements of the Merkle tree.
+func (m *MerkleTree) Leaves() []Leaf {
+	return m.leaves
+}
+
+// Root returns the root hash of the Merkle tree.
+func (m *MerkleTree) Root() []byte {
+	return m.tree[1].Hash()
+}
+
+// LeafCount returns the number of leaves in the Merkle tree.
+func (m *MerkleTree) LeafCount() int {
+	return len(m.leaves)
+}
+
+// BuildTree builds a serialized Merkle tree from an array of leaf nodes.
+//
+// The tree is 1-indexed, so root is at index 1.
+// The internal nodes are at index 2 to N.
+// The leaves are at index N+1 to 2N-1.
+//
+// For any node at index i:
+// - left child is at index 2*i
+// - right child is at index 2*i+1
+// - parent is at floor(i/2)
+func makeTree(nodes []Node) ([]Node, error) {
+	if len(nodes) == 0 {
+		return nil, ErrTreeEmpty
+	}
+
+	balancedLeafCount, err := CalculateBalancedNodesCount(len(nodes))
+	if err != nil {
+		return nil, err
+	}
+
+	// Allocate 2N space for the tree. (N leaf nodes, N-1 internal nodes)
+	tree := make([]Node, balancedLeafCount<<1)
+
+	lowerBound := balancedLeafCount
+	upperBound := balancedLeafCount + len(nodes) - 1
+
+	// Copy leaves into the tree, starting at index N.
+	for i := 0; i < len(nodes); i++ {
+		tree[lowerBound+i] = nodes[i]
+	}
+
+	for i := len(nodes); i < balancedLeafCount; i++ {
+		tree[lowerBound+i] = Node{hash: HashEmptyLeaf()}
+	}
+
+	// Build the tree.
+	for i := balancedLeafCount - 1; i >= 1; i-- {
+		leftChildIndex := GetLeftChild(i)
+
+		if leftChildIndex > upperBound {
+			continue
+		}
+
+		// Detect the level is processed.
+		if leftChildIndex <= lowerBound {
+			lowerBound >>= 1
+			upperBound >>= 1
+		}
+
+		rightChildIndex := GetRightChild(i)
+
+		tree[i] = Node{
+			hash: HashNode(tree[leftChildIndex].Hash(), tree[rightChildIndex].Hash()),
+		}
+
+	}
+
+	return tree, nil
+}
+
+// makeNodes returns the hashed leaves of the tree,
+// ordered in the same order as the provided elements.
+func makeNodes(leaves []Leaf) ([]Node, error) {
+	if len(leaves) == 0 {
+		return nil, ErrNoLeaves
+	}
+
+	nodes := make([]Node, len(leaves))
+	for i, element := range leaves {
+		nodes[i] = Node{hash: HashLeaf(element)}
+	}
+
+	return nodes, nil
+}
+
+// makeLeaves creates a balanced leaves count for the given leaves.
+// It fills empty leaves with an empty leaf hash.
+func makeLeaves(leaves []Leaf) ([]Leaf, error) {
+	balancedLeafCount, err := CalculateBalancedNodesCount(len(leaves))
+	if err != nil {
+		return nil, err
+	}
+
+	balancedLeaves := make([]Leaf, balancedLeafCount)
+
+	copy(balancedLeaves, leaves)
+
+	for i := len(leaves); i < balancedLeafCount; i++ {
+		balancedLeaves[i] = []byte{}
+	}
+
+	return balancedLeaves, nil
+}
+
+// CalculateBalancedNodesCount returns the number of nodes in a balanced tree.
+// To calculate the number of nodes in a tree, we need to round up to the next power of 2.
+// Returns an error if the element count is too large to be represented in a uint32.
```

Comment:
these type casts here... It's such a shame that Go `len()` is `int` instead of `uint`. I understand the design decision, but it makes some things a bit more annoying to cast.

I would assume that all of our code will be running on 64-bit architecture and so int will be int64. `roundUpToPowerOf2` can be easily adapted to accept and return `uint64`. Just add a few more shifts :)

That might remove the need for the overflow check and a bunch of these casts.

Reply:
```
func roundUpToPowerOf2(n int) int {
	if n <= 0 {
		return 1 // Smallest valid power of 2
	}

	// Use appropriate width based on architecture
	if bits.UintSize == 32 {
		v := uint32(n)
		if bits.OnesCount32(v) == 1 {
			return n
		}
		v |= v >> 1
		v |= v >> 2
		v |= v >> 4
		v |= v >> 8
		v |= v >> 16
		return int(v + 1)
	} else {
		v := uint64(n)
		if bits.OnesCount64(v) == 1 {
			return n
		}
		v |= v >> 1
		v |= v >> 2
		v |= v >> 4
		v |= v >> 8
		v |= v >> 16
		v |= v >> 32
		return int(v + 1)
	}
}
```

Reply:
> I would assume that all of our code will be running on 64-bit architecture and so int will be int64. roundUpToPowerOf2 can be easily adapted to accept and return uint64. Just add a few more shifts :) Or make it accept int and handle both architectures.

My assumption too, extracting this to an issue to work on it.